### PR TITLE
Use utf8mb4 charset to support emoji

### LINKED ISSUE #11083: Add Emoji Support
Actually with last version of PS1.7 it's still not possible to add emojis to products and categories, apart from the few supported by UTF-8 encoding.
I think in 2018 it's important having the possibility to add emoticon, in particular on meta descriptions and meta titles, because Google it's showing them also on SERP.

To do this it's needed to change utf8 encoding to utf8mb4:
https://mathiasbynens.be/notes/mysql-utf8mb4

### PR DESCRIPTION
**Pull request by PierreRambaud**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Use utf8mb4 to allow emoji. Also add the right capability to manage environment (dev, test, and prod)
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Ticket? | Fixes #11083
| How to test?  | You're able to upgrade, install and add emoji in text. :warning: The index size for utf8mb4 fields will changes to 191 due to a limitations with old MariaDb / Mysql instances.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/14894)

### CODE REVIEW

File: app/config/config.yml [hunk 1]
```diff
@@ -83,7 +83,7 @@ doctrine:
                 user:     "%database_user%"
                 password: "%database_password%"
                 server_version: 5.1
```
Comment (Progi1984) [inline app/config/config.yml]: MySQL 5.1 doesn't support utf8mb4. Your MySQL server must have a version 5.5.3 or higher (including all 10.x versions of MariaDB).

Comment (PierreRambaud) [reply app/config/config.yml]: trolololol
> Always wrap the server version number with quotes to parse it as a string instead of a float number. Otherwise, the floating-point representation issues can make your version be considered a different number (e.g. 5.6 will be rounded as 5.5999999999999996447286321199499070644378662109375).

I remove this configuration key because we can't specifcy which version the merchant is running. Nice catch!

Comment (Progi1984) [reply app/config/config.yml]: :+1:

**Review by Progi1984 — DISMISSED** [review]

File: composer.json [hunk 2]
```diff
@@ -153,8 +153,7 @@
             "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
         ],
         "create-test-db": [
-            "@php ./tests-legacy/create-test-db.php",
-            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache"
```
Comment (PierreRambaud) [inline composer.json]: is removed because not needed and not taking in consideration the current environment

File: .travis.yml [hunk 3]
```diff
@@ -72,7 +72,7 @@ before_install:
   # PrestaShop configuration
   - cp tests-legacy/parameters.yml.travis app/config/parameters.yml
 
-before_script:
+install:
```
Comment (Progi1984) [inline .travis.yml:79]: Why have you changed that ?

File: composer.json (hunk 2 above)
Comment (Progi1984) [inline composer.json]: Why did you remove it ?

File: config/defines.inc.php [hunk 4]
```diff
@@ -68,10 +68,30 @@
 if (defined('_PS_ADMIN_DIR_')) {
     define('_PS_BO_ALL_THEMES_DIR_', _PS_ADMIN_DIR_.'/themes/');
 }
+
+// Find if we are running under a Symfony command
+$cliEnvValue = null;
+if (isset($argv) && is_array($argv)) {
+    if (in_array('--env', $argv)) {
```
Comment (PierreRambaud) [inline config/defines.inc.php]: Because Symfony is not available.. No autoloader loaded. :trollface:

File: src/PrestaShopBundle/Service/Database/Upgrade.php [hunk 5]
```diff
@@ -39,8 +39,9 @@ public function addDoctrineSchemaUpdate()
         $command = new UpdateSchemaCommand();
         $this->application->add($command);
 
-        $this->commands[] = array(
+        $this->commands[] = [
             'command' => 'prestashop:schema:update-without-foreign',
-        );
+            '--env' => _PS_ENV_,
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Service/Database/Upgrade.php]: Make sure environement is the same when running subcommand.

File: src/PrestaShopBundle/Command/UpdateSchemaCommand.php [hunk 6]
```diff
@@ -58,16 +58,12 @@ protected function configure()
...
-            $this->dbName = $config['parameters']['database_name'];
-        }
+        $container = $this->getContainer();
 
-        $this->dbPrefix = $config['parameters']['database_prefix'];
+        $this->dbName = $container->getParameter('database_name');
+        $this->dbPrefix = $container->getParameter('database_prefix');
 
-        $this->em = $this->getContainer()->get('doctrine')->getManager();
+        $this->em = $container->get('doctrine')->getManager();
```
Comment (Progi1984) [inline src/PrestaShopBundle/Command/UpdateSchemaCommand.php]: You can use DI the EntityManager

**Review by Progi1984 — CHANGES_REQUESTED** [review]

File: src/PrestaShopBundle/Command/UpdateSchemaCommand.php (hunk 6 above)
Comment (PierreRambaud) [reply src/PrestaShopBundle/Command/UpdateSchemaCommand.php]: What do you mean?

File: .travis.yml (hunk 3 above)
Comment (PierreRambaud) [reply .travis.yml:79]: Because dependencies must be installed in travis' `install` step

File: composer.json (hunk 2 above)
Comment (PierreRambaud) [reply composer.json]: https://github.com/PrestaShop/PrestaShop/pull/14894/files#r310165191

File: .travis.yml (hunk 3 above)
Comment (Progi1984) [reply .travis.yml:79]: :+1:

File: src/PrestaShopBundle/Command/UpdateSchemaCommand.php (hunk 6 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Command/UpdateSchemaCommand.php]: EntityManager is a dependency. You can inject it in the constructor ?

Comment (PierreRambaud) [reply src/PrestaShopBundle/Command/UpdateSchemaCommand.php]: Not as I know because of this: https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Service/Database/Upgrade.php#L39 we don't have the container here :/
And the command already play with `getContainer`

Comment (Progi1984) [reply src/PrestaShopBundle/Command/UpdateSchemaCommand.php]: :ok_hand:  for me :)

**Review by Progi1984 — DISMISSED** [review]

File: classes/PrestaShopAutoload.php [hunk 7]
```diff
@@ -32,7 +32,7 @@
 class PrestaShopAutoload
 {
     /**
-     * @var PrestaShopAutoload
+     * @var PrvvestaShopAutoload
```
Comment (PierreRambaud) [inline classes/PrestaShopAutoload.php]:
```suggestion
     * @var PrestaShopAutoload
```

**Review by Progi1984 — DISMISSED** [review]

**Comment by PierreRambaud** [comment]
I add a WIP statement to split this PR.
One for the env management
One for utfmb4

**Comment by PierreRambaud** [comment]
WIP require #15902 and after I will rebase and squash the PR

**Comment by PierreRambaud** [comment]
No more in WIP

**Review by Progi1984 — APPROVED** [review]

**Comment by matks** [comment]
Thanks @PierreRambaud and congrats for this long-term feature !

**Comment by 1lymax** [comment]
How can I upgrade this over already installed Prestashop on my server? I can change my php/yml files, but how can I change .sql files? also I don't have install-dev directory.
Thanks.

**Comment by PierreRambaud** [comment]
@1lymax You need to wait for the 1.7.7.0 release :wink:

**Comment by NicoGag** [comment]
Thx a lot @PierreRambaud

**Comment by srnatan** [comment]
Thanks @PierreRambaud works in 1.7.6.9
